### PR TITLE
feat(authz): Add custom role checks in authorization

### LINKED ISSUE #3717: feat: change role apis to support custom roles

### LINKED ISSUE #3718: feat: integrate custom roles with authz

### PR DESCRIPTION
**Pull request by ThisIsMani**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR will add checks in authorization to check for custom roles.
This PR also add custom roles functionality to apis which were dealing with roles.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

To support custom roles.
Closes #3717 
Closes #3718 

### Roles
| Group | Org Admin | Merchant Admin | Merchant View Only | Merchant IAM Admin | Merchant Developer | Merchant Operator | Merchant Customer Support |
| ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- |
| Operations View | &check; | &check; | &check; | &check; | &check; | &check; | &check; |
| Operations Manage | &check; | &check; |  |  |  | &check; |  |
| Connectors View | &check; | &check; | &check; |  | &check; | &check; |  |
| Connectors Manage | &check; | &check; |  |  |  |  |  |
| Workflows View | &check; | &check; | &check; |  |  | &check; |  |
| Workflows Manage | &check; | &check; |  |  |  |  |  |
| Analytics View | &check; | &check; | &check; | &check; | &check; | &check; | &check; |
| Users View | &check; | &check; | &check; | &check; | &check; | &check; | &check; |
| Users Manage | &check; | &check; |  | &check; |  |  |  |
| Merchant Details View | &check; | &check; | &check; | &check; | &check; | &check; | &check; |
| MerchantDetails Manage | &check; | &check; |  |  | &check; |  |  |
| Organization Manage | &check; |  |  |  |  |  |  |

### Groups
| Permission | Operations View | Operations Manage | Connectors View | Connectors Manage | Workflows View | Workflows Manage | Analytics View | Users View | Users Manage | Merchant Details View | Merchant Details Manage | Organization Manage |
| ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- |
| Payment Read | &check; |  |  |  |  |  |  |  |  |  |  |  |
| Payment Write |  | &check; |  |  |  |  |  |  |  |  |  |  |
| Refund Read | &check; |  |  |  |  |  |  |  |  |  |  |  |
| Refund Write |  | &check; |  |  |  |  |  |  |  |  |  |  |
| Dispute Read | &check; |  |  |  |  |  |  |  |  |  |  |  |
| Dispute Write |  | &check; |  |  |  |  |  |  |  |  |  |  |
| Mandate Read | &check; |  |  |  |  |  |  |  |  |  |  |  |
| Mandate Write |  | &check; |  |  |  |  |  |  |  |  |  |  |
| Customer Read | &check; |  |  |  |  |  |  |  |  |  |  |  |
| Customer Write |  | &check; |  |  |  |  |  |  |  |  |  |  |
| Api Key Read |  |  |  |  |  |  |  |  |  |  | &check; |  |
| Api Key Write |  |  |  |  |  |  |  |  |  |  | &check; |  |
| Merchant Account Read | &check; | &check; | &check; | &check; | &check; | &check; | &check; | &check; | &check; | &check; | &check; | &check; |
| Merchant Account Write |  |  |  |  |  |  |  |  |  |  | &check; |  |
| Merchant Connector Account Read |  |  | &check; |  | &check; | &check; |  |  |  |  |  |  |
| Merchant Connector Account Write |  |  |  | &check; |  |  |  |  |  |  |  |  |
| Routing Read |  |  |  |  | &check; |  |  |  |  |  |  |  |
| Routing Write |  |  |  |  |  | &check; |  |  |  |  |  |  |
| Analytics |  |  |  |  |  |  | &check; |  |  |  |  |  |
| 3DS Manager Read |  |  |  |  | &check; |  |  |  |  |  |  |  |
| 3DS Manger Write |  |  |  |  |  | &check; |  |  |  |  |  |  |
| Surcharge Decision Manager Read |  |  |  |  | &check; |  |  |  |  |  |  |  |
| Surcharge Decision Manager Write |  |  |  |  |  | &check; |  |  |  |  |  |  |
| Users Read |  |  |  |  |  |  |  | &check; |  |  |  |  |
| Users Write |  |  |  |  |  |  |  |  | &check; |  |  |  |
| Merchant Account Create |  |  |  |  |  |  |  |  |  |  |  | &check; |

All the APIs will perform according to these roles from now on.
----
```
curl --location 'http://localhost:8080/user/role/list' \
--header 'Authorization: Bearer JWT'
```
```
[
    {
        "role_id": "merchant_developer",
        "permissions": [
            "CustomerRead",
            "MerchantAccountRead",
            "Analytics",
            "UsersRead",
            "PaymentRead",
            "MerchantAccountWrite",
            "MerchantConnectorAccountRead",
            "ApiKeyRead",
            "MandateRead",
            "RefundRead",
            "DisputeRead",
            "ApiKeyWrite"
        ],
        "role_name": "Developer",
        "role_scope": "organization"
    },
    {
        "role_id": "merchant_operator",
        "permissions": [
            "DisputeWrite",
            "RefundWrite",
            "SurchargeDecisionManagerRead",
            "CustomerWrite",
            "DisputeRead",
            "MandateWrite",
            "MerchantConnectorAccountRead",
            "MerchantAccountRead",
            "PaymentWrite",
            "MandateRead",
            "RefundRead",
            "PaymentRead",
            "RoutingRead",
            "ThreeDsDecisionManagerRead",
            "Analytics",
            "UsersRead",
            "CustomerRead"
        ],
        "role_name": "Operator",
        "role_scope": "organization"
    },
    {
        "role_id": "merchant_iam_admin",
        "permissions": [
            "PaymentRead",
            "MerchantAccountRead",
            "Analytics",
            "DisputeRead",
            "UsersRead",
            "MandateRead",
            "UsersWrite",
            "RefundRead",
            "CustomerRead"
        ],
        "role_name": "IAM",
        "role_scope": "organization"
    },
    {
        "role_id": "merchant_view_only",
        "permissions": [
            "MerchantConnectorAccountRead",
            "SurchargeDecisionManagerRead",
            "UsersRead",
            "Analytics",
            "ThreeDsDecisionManagerRead",
            "MandateRead",
            "DisputeRead",
            "MerchantAccountRead",
            "RoutingRead",
            "CustomerRead",
            "RefundRead",
            "PaymentRead"
        ],
        "role_name": "View Only",
        "role_scope": "organization"
    },
    {
        "role_id": "merchant_customer_support",
        "permissions": [
            "RefundRead",
            "UsersRead",
            "Analytics",
            "CustomerRead",
            "MandateRead",
            "DisputeRead",
            "MerchantAccountRead",
            "PaymentRead"
        ],
        "role_name": "Customer Support",
        "role_scope": "organization"
    },
    {
        "role_id": "merchant_admin",
        "permissions": [
            "Analytics",
            "RoutingRead",
            "ApiKeyRead",
            "MerchantAccountRead",
            "MandateWrite",
            "RefundWrite",
            "SurchargeDecisionManagerRead",
            "ThreeDsDecisionManagerWrite",
            "CustomerRead",
            "MerchantAccountWrite",
            "ApiKeyWrite",
            "ThreeDsDecisionManagerRead",
            "PaymentWrite",
            "MerchantConnectorAccountRead",
            "PaymentRead",
            "MerchantConnectorAccountWrite",
            "UsersRead",
            "UsersWrite",
            "DisputeRead",
            "MandateRead",
            "DisputeWrite",
            "CustomerWrite",
            "RoutingWrite",
            "RefundRead",
            "SurchargeDecisionManagerWrite"
        ],
        "role_name": "Admin",
        "role_scope": "organization"
    }
]
```
---

```
curl --location 'http://localhost:8080/user/role/role_id' \
--header 'Authorization: Bearer JWT'
```

```
{
    "role_id": "rold_id",
    "permissions": [ permissions as mentioned above ],
    "role_name": "name",
    "role_scope": "organization"
}
```
## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/services/authorization/permissions.rs [hunk 1]
```diff
@@ -1,6 +1,6 @@
 use strum::Display;
 
-#[derive(PartialEq, Display, Clone, Debug, Copy)]
+#[derive(PartialEq, Display, Clone, Debug, Copy, Eq, Hash)]
```
Comment (racnan) [inline crates/router/src/services/authorization/permissions.rs:3]: why is `hash` required ?

File: crates/router/src/services/authorization/roles.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
+use common_enums::{PermissionGroup, RoleScope};
+use common_utils::errors::CustomResult;
+
+use super::{permission_groups::get_permissions_vec, permissions::Permission};
+use crate::{core::errors, routes::AppState};
+
+pub mod predefined_roles;
+
+#[allow(dead_code)]
```
Comment (racnan) [inline crates/router/src/services/authorization/roles.rs:9]: why is this required ?

File: crates/router/src/services/authorization/roles.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    state: &AppState,
+    role_id: &str,
+) -> CustomResult<RoleInfo, errors::StorageError> {
+    if let Some(role) = predefined_roles::PREDEFINED_ROLES.get(role_id) {
+        Ok(role.clone())
+    } else {
+        let role = state.store.find_role_by_role_id(role_id).await?;
+        Ok(role.into())
+    }
+}
```
Comment (racnan) [inline crates/router/src/services/authorization/roles.rs:79]: re-evaluate use of the function as it allows you to get any custom role irrespective of the merchant id.

**Review by racnan — CHANGES_REQUESTED** [review]

File: crates/router/src/routes/files.rs [hunk 4]
```diff
@@ -47,7 +47,7 @@ pub async fn files_create(
         |state, auth, req| files_create_core(state, auth.merchant_account, auth.key_store, req),
         auth::auth_type(
             &auth::ApiKeyAuth,
-            &auth::JWTAuth(Permission::FileWrite),
+            &auth::DashboardNoPermissionAuth,
```
[the remark this answers is not in the document]
Comment (ThisIsMani) [reply crates/router/src/routes/files.rs:50]: Product confirmed that these apis should have JWT.

**Comment by racnan** [comment]
add test cases.

**Review by ShivanshMathurJuspay — APPROVED** [review]

**Pull request merged**